src/controller/AutoCommissioner.cpp

```diff
@@ -392,6 +392,13 @@ CommissioningStage AutoCommissioner::GetNextCommissioningStageInternal(Commissio
         }
         return CommissioningStage::kFindOperational;
     case CommissioningStage::kFindOperational:
+        if (mParams.GetIcdRegistrationStrategy() != IcdRegistrationStrategy::kIgnore)
+        {
+            return CommissioningStage::kIcdIdentification;
+        }
+        return CommissioningStage::kSendComplete;
```

Comment:
This is not safe.  You _have_ to do operational session lookup right before doing SendComplete, to avoid failures.  If something else happened after the last operational lookup, you have to do it again.

Reply:
> what is the issue of having other steps between OpLookup & SendComplete?

While the OpLookup callbacks are firing, the CASE session is known to exist.  As soon as they return, you are no longer guaranteed that.

The fact that there is a persistent "device" object that pretends to keep the session alive in this case does not mean that it actually stays alive.

> we have to add ICD Registration call before SendComplete.

Sure.  It just needs to happen before the OpLookup step.

Let me ask a different question: why was it placed after the OpLookup step?

> What makes it non-safe for adding steps which does not modify the state of the session?

What makes you think nothing is modifying the state of the session?  Keep in mind that the session pool is shared across all controllers, etc.

> it looks safe for reusing the SessionHandle / device proxy, what makes the difference

It's safe to use the SessionHandle from an oplookup while inside the callbacks from the oplookup.  Any time after that you have absolutely no guarantees the session still exists.  DeviceProxy does not keep the session alive.

> Or the SendComplete stage should do an extra FindOrEstablish by itself.

That would be fine, yes, but that's not how this code is structured.

Just to reiterate: the only safe way to do things over CASE in our current architecture is to do what this code calls "OpLookup" (aka asking the CASESessionManager  for a session) to get the session before any creation of a new exchange.

Reply:
> Sure. It just needs to happen before the OpLookup step.
> 
> Let me ask a different question: why was it placed after the OpLookup step?

Well, just checked that after AddNOC, the PASE session has an associating fabric, so calling before OpLookup should be OK.

> > What makes it non-safe for adding steps which does not modify the state of the session?
> 
> What makes you think nothing is modifying the state of the session? Keep in mind that the session pool is shared across all controllers, etc.

The separate stage - OpLookup and SendComplete indicates that -- I believe this is an issue and should be resolved (in separate PR).

And we are not actually using the SessionHandle in the callback of OpLookup, at least the code structure indicates this.

Reply:
> Well, just checked that after AddNOC, the PASE session has an associating fabric

Right, this needs to come after AddNOC, but before OpLookup.  Where to put it with respect to network config, I am not sure.  If there are considerations to think about here and we decide on an ordering wrt network config based on those, we should document them.

Note that the code you had was doing the ICD bits over PASE anyway.  It was just doing it after CASE establishment for some reason.  See the implementation of `AutoCommissioner::GetDeviceProxyForStep`.

> And we are not actually using the SessionHandle in the callback of OpLookup

We _are_ it's just well-hidden.  In `DeviceCommissioner::OnDeviceConnectedFn` we do:
```
        report.Set<OperationalNodeFoundData>(OperationalNodeFoundData(OperationalDeviceProxy(&exchangeMgr, sessionHandle)));
        commissioner->CommissioningStageComplete(CHIP_NO_ERROR, report);
```
which is sending the `SessionHandle` to `CommissioningStageComplete`, as part of an ephemeral "device", because this code never got fully updated to the API changes to CASE establishment, so it's using this "device" thing that's deprecated as a shim.  That then turns around and synchronously calls into `AutoCommissioner::CommissioningStepFinished`, which does:
```
        case CommissioningStage::kFindOperational:
            mOperationalDeviceProxy = report.Get<OperationalNodeFoundData>().operationalProxy;
```
which is storing the still-ephemeral session info in `mOperationalDeviceProxy`.  Then it synchronously does `AutoCommissioner::PerformStep`, which calls `DeviceCommissioner::PerformCommissioningStep` which sends the `CommissioningComplete` command.  This all happens synchronously under the "we have a session" callback, which is why it's ok in the end.

Reply:
Moved this to ReadCommissioningInfo2